### PR TITLE
Support `map`

### PR DESCRIPTION
**Pull request by themattchan**
... and also fix a lot of stylistic nits in `Proto3.Suite.DotProto.Generate`.

The reference implementation of this feature is an ugly hack. The idea is to apply the following desugaring transformation:

```
message MyMessage {
  map<KTy, VTy> myMap = 1;
}

==>

message MyMessage {
  message MyMapEntry {
    KTy key = 1;
    VTy value = 2;
  }
  repeated MyMapEntry myMap = 1;
}
```

In Haskell you do not need to do this -- simply use the generic `Message` instance for tuples.

### CODE REVIEW

File: test-files/test_proto.proto [hunk 1]
```diff
@@ -139,4 +139,15 @@ message Wrapped {
 enum EnumAnnots {
   FOO = 0;
   BAR = 1 [deprecated=true];
+}
+
+message MapTest {
+  map<string, sint32> prim = 1;
+  map<int32, Wrapped> wrapped = 2;
+  // In theory you could do this because of how map is elaborated to a
+  // nested message under the hood. In practice it causes the reference impl to barf.
```
Comment (Gabriel439) [inline test-files/test_proto.proto]: Could you expand on this comment?  In particular, what issues do you see in the reference implementation?

Comment (themattchan) [reply test-files/test_proto.proto]: i thought i removed this...

File: src/Proto3/Suite/Class.hs [hunk 2]
```diff
@@ -469,13 +474,47 @@ instance MessageField B.ByteString
...
 instance (Bounded e, Named e, Enum e) => MessageField (Enumerated e)
 
+instance (Ord k, Primitive k, Primitive v) => MessageField (M.Map k v) where
+  encodeMessageField num = foldMap (\(k, v) -> Encode.embedded num $
+                                     encodePrimitive (fieldNumber 1) k <>
+                                     encodePrimitive (fieldNumber 2) v
+                                   )
+                           . M.toList
+
+  -- 'reverse' the list before turning into a map because the map monoid keeps
```
[the remark this answers is not in the document]
Comment (themattchan) [reply src/Proto3/Suite/Class.hs]: Removing `reverse`. Should I leave the comment in or remove it?

**Comment by themattchan** [comment]
This works now, and is ready for another review.

File: tests/TestCodeGen.hs [hunk 3]
```diff
@@ -77,35 +77,37 @@ setPythonPath = Turtle.export "PYTHONPATH" =<< do
 
 simpleEncodeDotProto :: TestTree
 simpleEncodeDotProto =
-    testCase "generate code for a simple .proto and then use it to encode messages" $
-    do compileTestDotProtos
-       -- Compile our generated encoder
-       (@?= ExitSuccess) =<< Turtle.proc "tests/encode.sh" [hsTmpDir] empty
+    testCase "generate code for a simple .proto and then use it to encode messages"
+    $ do
+         compileTestDotProtos
```
Comment (themattchan) [inline tests/TestCodeGen.hs]: I did this to make the `compileTestDotProtos` step more obvious

File: src/Proto3/Suite/DotProto/Generate.hs [hunk 4]
```diff
@@ -1387,97 +1378,102 @@ oneofSubDisjunctBinder = intercalate "_or_" . fmap oneofSubBinder
...
 
-wrapE
-    :: MonadError CompileError m
-    => TypeContext
-    -> DotProtoType
-    -> [DotProtoOption]
-    -> HsExp
-    -> m HsExp
-wrapE ctxt dpt opts e = case dpt of
+wrapE :: TypeContext -> DotProtoType -> [DotProtoOption] -> HsExp -> HsExp
```
Comment (themattchan) [inline src/Proto3/Suite/DotProto/Generate.hs]: Most of the changes in this file are just style/formatting changes. This block is where the action is.

**Pull request merged**